geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/data/KafkaCacheLoader.scala

```diff
@@ -113,5 +123,90 @@ object KafkaCacheLoader {
       }
     }
   }
+
+  /**
+    * Handles initial loaded 'from-beginning' without indexing features in the spatial index. Will still
+    * trigger message events.
+    *
+    * @param consumers consumers, won't be closed even on call to 'close()'
+    * @param topic kafka topic
+    * @param frequency polling frequency in milliseconds
+    * @param serializer message serializer
+    * @param toLoad main cache loader, used for callback when bulk loading is done
+    */
+  private class InitialLoader(override protected val consumers: Seq[Consumer[Array[Byte], Array[Byte]]],
+                              override protected val topic: String,
+                              override protected val frequency: Long,
+                              serializer: GeoMessageSerializer,
+                              toLoad: KafkaCacheLoaderImpl) extends ThreadedConsumer with Runnable {
+
+    private val loadCache: Cache[String, SimpleFeature] = Caffeine.newBuilder().build()
+
+    // track the offsets that we want to read to
+    private val offsets = new ConcurrentHashMap[Int, Long]()
+    private val done = new AtomicBoolean(false)
+
+    override protected def closeConsumers: Boolean = false
+
+    override protected def consume(record: ConsumerRecord[Array[Byte], Array[Byte]]): Unit = {
+      if (done.get) { toLoad.consume(record) } else {
+        val message = serializer.deserialize(record.key, record.value)
+        logger.trace(s"Consumed message [$topic:${record.partition}:${record.offset}] $message")
+        message match {
+          case m: Change => toLoad.fireEvent(m); loadCache.put(m.feature.getID, m.feature)
+          case m: Delete => toLoad.fireEvent(m); loadCache.invalidate(m.id)
+          case m: Clear  => toLoad.fireEvent(m); loadCache.invalidateAll()
+          case m => throw new IllegalArgumentException(s"Unknown message: $m")
+        }
+        // once we've hit the max offset for the partition, remove from the offset map to indicate we're done
+        if (offsets.getOrDefault(record.partition, Long.MaxValue) <= record.offset) {
+          offsets.remove(record.partition)
+        }
+      }
+    }
+
+    override def run(): Unit = {
+      import scala.collection.JavaConverters._
+
+      val partitions = consumers.head.partitionsFor(topic).asScala.map(_.partition)
```

Comment:
Sanity check:  each of the consumers has the same number of partitions?

Reply:
this is just a metadata operation for the topic - it doesn't necessarily mean that the consumer is assigned a particular partition. there is another method `assignment` that does that.